### PR TITLE
[Core] - Support azure-core version without OpenSSL dependency

### PR DESCRIPTION
**Pull request by vhvb1989**
An azure-core lib version without OpenSSL is required for the speech service which will consume only UUID (and HTTP in some future).

This PR makes some changes to the current azure core components consuming OpenSSL :

- Uuid on Linux :  It is used to generate a random buffer.  Switching to use the win-implementation which uses the std lib <random> lib.
- base64 :  Removing the current implementation calling openSSL for base64 for the azure-core for C99 implementation created in the past by @antkmsft 
- Crypto / Hashing :  MD5 and Sha hashing is currently exposed on the public surface and implemented by calling OpenSSL lib.  This PR is making all the Cryptographic namespace an optional feature that can be opt-out when building the azure-core lib

### CODE REVIEW

**Review by RickWinter — CHANGES_REQUESTED** [review]
This change will result in non-unique UUIDs on linux platforms.  std::random is not cryptographically secure on linux

File: sdk/core/azure-core/src/uuid.cpp [hunk 1]
```diff
@@ -45,24 +41,13 @@ namespace Azure { namespace Core {
   {
     uint8_t uuid[UuidSize] = {};
 
-#if defined(AZ_PLATFORM_WINDOWS)
     std::random_device rd;
```
Comment (RickWinter) [inline sdk/core/azure-core/src/uuid.cpp:44]: Using std::random_device on linux will result in UUIDs which are not unique

**Comment by vhvb1989** [comment]
64-bit Mersenne Twister by Matsumoto
=== Results ===
Completed 536,943 operations in a weighted-average of 10s (53,693.482313 ops/s, 1.86242e-05 s/op)

OpenSSL
=== Results ===
Completed 44,534 operations in a weighted-average of 10s (4,453.341869 ops/s, 0.00022455 s/op)

File: sdk/core/azure-core/src/base64.cpp [hunk 2]
```diff
@@ -4,93 +4,474 @@
...
+static int32_t base64Encode(const uint8_t* threeBytes)
+{
+  int32_t i = (*threeBytes << 16) | (*(threeBytes + 1) << 8) | *(threeBytes + 2);
+
+  int32_t i0 = Base64EncodeArray[i >> 18];
+  int32_t i1 = Base64EncodeArray[(i >> 12) & 0x3F];
+  int32_t i2 = Base64EncodeArray[(i >> 6) & 0x3F];
+  int32_t i3 = Base64EncodeArray[i & 0x3F];
+
+  return i0 | (i1 << 8) | (i2 << 16) | (i3 << 24);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:284]: This looked funnily similar. [Great idea!](https://github.com/Azure/azure-sdk-for-c/pull/1816/)

I am assuming going from az_span to pointers was relatively straightforward, right?

For reviewers, the key would be to look out for any out-of-bounds or integer overflow issues.

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,10 @@
 
 ### Other Changes
 
+- Added option `BUILD_CRYPTO` to support building azure-core without the Cryptographic namespace (no hashing). ON by default.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:13]: Can you help me understand why we want to expose this build option to the customer? What scenario requires disabling building this namespace and why?

File: sdk/core/azure-core/CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,10 @@
...
+- Updated `base64` implementation.
+- Updated `Uuid` implementation for Linux.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:15]: The reader of the CL might wonder what the update was or why it was made, otherwise, it isn't useful.

```suggestion
- Updated `base64` implementation to remove external dependency.
- Updated `Uuid` implementation for Linux to remove external dependency.
```

File: sdk/core/azure-core/CMakeLists.txt [hunk 5]
```diff
@@ -22,6 +22,10 @@ az_vcpkg_integrate()
 
 find_package(Threads REQUIRED)
 
+# Allow building azure-core without OpenSSL dependency and hence without Crypto-functions
+# This is a work-around for anyone consuming azure-core without a need for Crypto like the cognitive services.
+option(BUILD_CRYPTO "Build the cryto namespace. Build will depend on OpenSSL." ON)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:27]: I thought we are getting rid of the openssl dependency all together, no?

```suggestion
option(BUILD_CRYPTO "Build the crypto namespace. Build will depend on OpenSSL." ON)
```

File: sdk/core/azure-core/src/uuid.cpp [hunk 6]
```diff
@@ -45,24 +49,14 @@ namespace Azure { namespace Core {
   {
     uint8_t uuid[UuidSize] = {};
 
-#if defined(AZ_PLATFORM_WINDOWS)
-    std::random_device rd;
+    // RAND_MAX == INT_MAX
+    std::uniform_int_distribution<uint32_t> distribution(0, RAND_MAX);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:53]: Why are we asking for uint32_t but then using INT_MAX instead? Should this be UINT_MAX?

**Review by ahsonkhan — COMMENTED** [review]
I am not sure we should expose the build crypto option, if we can remove the openssl dependency all together.

Will review base64 impl in a bit, do we have sufficient test coverage for it?

File: sdk/core/azure-core/src/uuid.cpp [hunk 7]
```diff
@@ -45,24 +49,14 @@ namespace Azure { namespace Core {
...
 
     for (size_t i = 0; i < UuidSize; i += 4)
     {
-      const uint32_t x = rd();
+      const uint32_t x = distribution(randomGenerator);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/uuid.cpp:57]: How did we test the impl to increase confidence this provides unique uuids?

File: sdk/core/azure-core/CHANGELOG.md (hunk 3 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/CHANGELOG.md:13]: > Crypto / Hashing : MD5 and Sha hashing is currently exposed on the public surface and implemented by calling OpenSSL lib. This PR is making all the Cryptographic namespace an optional feature that can be opt-out when building the azure-core lib

Oh... it's because of hashing :(
We got rid of it from base64 and uuid, but we still have one component left

File: sdk/core/azure-core/src/base64.cpp [hunk 8]
```diff
@@ -4,93 +4,474 @@
...
-
 #include <string>
 #include <vector>
 
-namespace Azure { namespace Core {
+namespace {
+
+static char const Base64EncodeArray[65]
+    = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static char const EncodingPath = '=';
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:14]:
```suggestion
static char const EncodingPad = '=';
```

File: sdk/core/azure-core/src/base64.cpp [hunk 9]
```diff
@@ -4,93 +4,474 @@
...
+  *(destination + 1) = static_cast<uint8_t>((value >> 8) & 0xFF);
+  *(destination + 0) = static_cast<uint8_t>(value & 0xFF);
+}
+
+std::string base64Encode(const std::vector<uint8_t>& data)
+{
+  int64_t sourceIndex = 0;
+  auto inputSize = data.size();
+  int32_t result = 0;
+  auto maxEncodedSize = ((inputSize + 2) / 3) * 4;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:321]: This could overflow int32 if `data.size()` is larger than 1610612733. What do we do there?

File: sdk/core/azure-core/src/base64.cpp [hunk 10]
```diff
@@ -4,93 +4,474 @@
...
+  // Use a string with size to the max possible result
+  std::string encodedResult(maxEncodedSize, '0');
+  // Removing const from the string to update the placeholder string
+  auto destination = const_cast<char*>(encodedResult.data());
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:325]: I wonder if we should use some sort of string builder like stringstream instead of doing this.

@antkmsft, @CaseyCarter - what are your thoughts here?

File: sdk/core/azure-core/src/base64.cpp (hunk 9 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:321]: We may want to check and throw up front for such inputs, unless int64_t is the expected size we want to support and use. vector.size is size_t

File: sdk/core/azure-core/src/base64.cpp [hunk 11]
```diff
@@ -4,93 +4,474 @@
...
+    base64WriteIntAsFourBytes(destination, result);
+    destination += 4;
+    sourceIndex += 2;
+  }
+  auto destinationLength = static_cast<size_t>(destination - encodedResult.data());
+  // If encoding took less than the max-expected
+  if (destinationLength < encodedResult.size())
+  {
+    return encodedResult.substr(0, destinationLength);
+  }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:354]: Do we have a test case that exercises this code path?

File: sdk/core/azure-core/src/base64.cpp [hunk 12]
```diff
@@ -4,93 +4,474 @@
...
+}
+
+static void base64WriteThreeLowOrderBytes(std::vector<uint8_t>::iterator destination, int64_t value)
+{
+  *destination = static_cast<uint8_t>(value >> 16);
+  *(destination + 1) = static_cast<uint8_t>(value >> 8);
+  *(destination + 2) = static_cast<uint8_t>(value);
+}
+
+std::vector<uint8_t> base64Decode(const std::string& text)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:388]: Don't we use PascalCase in C++? Same issue elsewhere
```suggestion
std::vector<uint8_t> Base64Decode(const std::string& text)
```

File: sdk/core/azure-core/src/base64.cpp [hunk 13]
```diff
@@ -4,93 +4,474 @@
...
-
-    CryptBinaryToStringA(
-        reinterpret_cast<const BYTE*>(data.data()),
-        static_cast<DWORD>(data.size()),
-        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF,
-        static_cast<LPSTR>(&encoded[0]),
-        &encodedLength);
-
-    return encoded;
+    return std::vector<uint8_t>(0);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:393]: A non-empty input that is smaller than 4 characters, is invalid, and we should throw here, not return an empty result.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:393]: Also, the input must be a multiple of 4, so check for that and throw otherwise.

File: sdk/core/azure-core/src/base64.cpp [hunk 14]
```diff
@@ -4,93 +4,474 @@
...
+  auto inputSize = text.size();
+  auto inputPtr = text.data();
+  // use the size for the max decoded size
+  auto maxDecodedSize = (inputSize / 4) * 3;
+  std::vector<uint8_t> destination(maxDecodedSize);
+  auto destinationPtr = destination.begin();
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 4)
+  {
+    int64_t result = base64Decode(inputPtr + sourceIndex);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:407]: If `result` is negative, it means invalid input and we should throw.

We need test cases like the following, which are invalid input:
"A" - too small
"AB" - too small
"ABC" - too small
"ABCD==" - not a multiple of 4
"ABCD====" - invalid padding count
"A%%%" - non-base64 chars

Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:407]: The input has to be of characters from `"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"`

File: sdk/core/azure-core/src/base64.cpp [hunk 15]
```diff
@@ -4,93 +4,474 @@
...
-    return decoded;
+    i2 = Base64DecodeArray[i2];
+    i3 = Base64DecodeArray[i3];
+
+    i2 <<= 6;
+
+    i0 |= i3;
+    i0 |= i2;
+
+    base64WriteThreeLowOrderBytes(destinationPtr, i0);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:439]: Same here and elsewhere - add input validation checks (if i0 < 0) and throw for invalid input

https://github.com/Azure/azure-sdk-for-c/blob/d7b388c68b07ee769dd9b061114b93336fd53385/sdk/src/azure/core/az_base64.c#L487-L490

File: sdk/core/azure-core/src/base64.cpp [hunk 16]
```diff
@@ -4,93 +4,474 @@
...
-    BIO* bio = BIO_new_mem_buf(text.data(), -1);
-    bio = BIO_push(BIO_new(BIO_f_base64()), bio);
-    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
-    int decodedLength = BIO_read(bio, &decoded[0], static_cast<int>(text.length()));
-    BIO_free_all(bio);
-
-    decoded.resize(decodedLength);
-    return decoded;
+    destination.resize(resultSize);
+    destination.shrink_to_fit();
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/base64.cpp:464]: Resizing strings and vectors after the fact seems to be a relatively expensive operation.
@CaseyCarter how do we avoid this, and maybe just create the vector of the right size at the end without having to do linear data copy or re-allocation? In .NET/C#, this was relatively easy with span/slicing for buffers.

**Review by ahsonkhan — COMMENTED** [review]
I think we need more testing of the base64 implementation, especially around edge cases (for integer overflow) and invalid input.

File: sdk/core/azure-core/src/base64.cpp (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:354]: Doesn't look like it:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=1113355&view=codecoverage-tab
image

Coverable lines:	120
Total lines:	477
Line coverage:	99.1% (119 of 120)
Covered branches:	22
Total branches:	30
Branch coverage:	73.3% (22 of 30)

File: sdk/core/azure-core/src/base64.cpp [hunk 17]
```diff
@@ -4,93 +4,474 @@
...
+  auto maxDecodedSize = (inputSize / 4) * 3;
+  std::vector<uint8_t> destination(maxDecodedSize);
+  auto destinationPtr = destination.begin();
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 4)
+  {
+    int64_t result = base64Decode(inputPtr + sourceIndex);
+    base64WriteThreeLowOrderBytes(destinationPtr, result);
+    destinationPtr += 3;
+    destinationIndex += 3;
```
Comment (CaseyCarter) [inline sdk/core/azure-core/src/base64.cpp:410]: `destinationIndex` is unused.

File: sdk/core/azure-core/src/base64.cpp [hunk 18]
```diff
@@ -4,93 +4,474 @@
...
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1, // Bytes over 122 ('z') are invalid and cannot be decoded. Hence, padding the map with 255,
```
Comment (CaseyCarter) [inline sdk/core/azure-core/src/base64.cpp:159]:
```suggestion
    -1, // Bytes over 122 ('z') are invalid and cannot be decoded. Hence, padding the map with -1,
```

File: sdk/core/azure-core/src/base64.cpp [hunk 19]
```diff
@@ -4,93 +4,474 @@
...
-#if defined(AZ_PLATFORM_WINDOWS)
+static int64_t base64Decode(const char* encodedBytes)
+{
+  int64_t i0 = *encodedBytes;
+  int64_t i1 = *(encodedBytes + 1);
+  int64_t i2 = *(encodedBytes + 2);
+  int64_t i3 = *(encodedBytes + 3);
 
-  std::string Convert::Base64Encode(const std::vector<uint8_t>& data)
+  i0 = Base64DecodeArray[i0];
```
Comment (CaseyCarter) [inline sdk/core/azure-core/src/base64.cpp:365]: On platforms where `char` is signed (C and C++ let each implementation decide if `char` is signed or not) this could be accessing memory _before_ `Base64DecodeArray`. I recommend:
```c++
int32_t i0 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[0])];
int32_t i1 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[1])];
int32_t i2 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[2])];
int32_t i3 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[3])];
int32_t errcheck = (i0 | i1 | i2 | i3) & 0xC0); // If either of the high two bits are set in any result, the input is invalid
return (errcheck << 24) | i0 << 18 | i1 << 12 | i2 << 6 | i3; // If any of the high 8 bits is set, the result is invalid
```

File: sdk/core/azure-core/test/perf/inc/azure/core/test/uuid_test.hpp [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+  public:
+    /**
+     * @brief Construct a new Uuid test.
+     *
+     * @param options The test options.
+     */
+    UuidTest(Azure::Perf::TestOptions options) : PerfTest(options) {}
+
+    /**
+     * @brief Use Uuid to assing and read.
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/test/perf/inc/azure/core/test/uuid_test.hpp:32]:
```suggestion
     * @brief Use Uuid to assign and read.
```

File: sdk/core/azure-core/src/base64.cpp [hunk 21]
```diff
@@ -4,93 +4,474 @@
...
+  int64_t sourceIndex = 0;
+  auto inputSize = data.size();
+  int32_t result = 0;
+  auto maxEncodedSize = ((inputSize + 2) / 3) * 4;
+  // Use a string with size to the max possible result
+  std::string encodedResult(maxEncodedSize, '0');
+  // Removing const from the string to update the placeholder string
+  auto destination = const_cast<char*>(encodedResult.data());
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 2)
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:327]: Is `sourceIndex + 3` guaranteed not to overflow, once it is a `size_t`?

File: sdk/core/azure-core/src/base64.cpp [hunk 22]
```diff
@@ -4,93 +4,474 @@
...
+    return std::vector<uint8_t>(0);
   }
 
-  std::vector<uint8_t> Convert::Base64Decode(const std::string& text)
+  int64_t sourceIndex = 0;
+  int64_t destinationIndex = 0;
+  auto inputSize = text.size();
+  auto inputPtr = text.data();
+  // use the size for the max decoded size
+  auto maxDecodedSize = (inputSize / 4) * 3;
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:401]: We should be doing validation.

Also, we can't make it exact without actually counting the number of `=` at the end up front (to see if there is 0, 1, or 2 `=`), which is a bit unnecessary. The validation during the decoding pass should catch it.

File: sdk/core/azure-core/src/base64.cpp (hunk 19 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:365]: > C and C++ let each implementation decide if `char` is signed or not

🤦‍♂️

Why?! Lol

File: sdk/core/azure-core/src/base64.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply sdk/core/azure-core/src/base64.cpp:401]: > Also, we can't make it exact without actually counting the number of `=` at the end up front (to see if there is 0, 1, or 2 `=`), which is a bit unnecessary. The validation during the decoding pass should catch it.

I suppose I don't care whether we burn a dozen clock cycles to get the size exact or we burn a couple of unused bytes at the end of the vector that holds the decided result. I _do_ think it's a horrible idea to `shrink_to_fit` at the end, copying the entire decoded content into a freshly-allocated buffer, to avoid wasting said "couple of bytes."

File: sdk/core/azure-core/src/base64.cpp (hunk 19 above)
Comment (CaseyCarter) [reply sdk/core/azure-core/src/base64.cpp:365]: Pre-standardization implementation divergence, to my understanding.

File: sdk/core/azure-core/src/base64.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:401]: Good point! Agreed :)

File: sdk/core/azure-core/src/base64.cpp [hunk 23]
```diff
@@ -4,93 +4,474 @@
...
+  std::vector<uint8_t> destination(maxDecodedSize);
+  auto destinationPtr = destination.begin();
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 4)
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:405]: Is the `+ 4` safe from an integer overflow perspective?

File: sdk/core/azure-core/src/base64.cpp (hunk 19 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/base64.cpp:365]: Good thing the C impl is using `int32_t` then instead of char:
https://github.com/Azure/azure-sdk-for-c/blob/7f689bea833a12b0be9059590e7321d5af99296f/sdk/src/azure/core/az_base64.c#L380-L383

Though, now I wonder what invalid assumptions I made about C types when writing that code and whether that code could also benefit from another set of eyes to review :)
